packages/errors/src/types.ts

```diff
@@ -0,0 +1,56 @@
+import type { InvalidPayloadErrorExtensions } from './errors/invalid-payload.js';
+import type { InvalidProviderConfigErrorExtensions } from './errors/invalid-provider-config.js';
+import type { InvalidQueryErrorExtensions } from './errors/invalid-query.js';
+import type { MethodNotAllowedErrorExtensions } from './errors/method-not-allowed.js';
+import type { NotNullViolationErrorExtensions } from './errors/not-null-violation.js';
+import type { RangeNotSatisfiableErrorExtensions } from './errors/range-not-satisfiable.js';
+import type { RecordNotUniqueErrorExtensions } from './errors/record-not-unique.js';
+import type { RouteNotFoundErrorExtensions } from './errors/route-not-found.js';
+import type { ServiceUnavailableErrorExtensions } from './errors/service-unavailable.js';
+import type { UnprocessableContentErrorExtensions } from './errors/unprocessable-content.js';
+import type { UnsupportedMediaTypeErrorExtensions } from './errors/unsupported-media-type.js';
+import type { ValueOutOfRangeErrorExtensions } from './errors/value-out-of-range.js';
+import type { ValueTooLongErrorExtensions } from './errors/value-too-long.js';
+
+type Map = {
+	[ErrorCode.ContainsNullValues]: ContainsNullValuesErrorExtensions;
+	[ErrorCode.ContentTooLarge]: never;
+	[ErrorCode.Forbidden]: never;
+	[ErrorCode.IllegalAssetTransformation]: IllegalAssetTransformationErrorExtensions;
+	[ErrorCode.InvalidCredentials]: never;
+	[ErrorCode.InvalidForeignKey]: InvalidForeignKeyErrorExtensions;
+	[ErrorCode.InvalidIp]: never;
+	[ErrorCode.InvalidOtp]: never;
+	[ErrorCode.InvalidPayload]: InvalidPayloadErrorExtensions;
+	[ErrorCode.InvalidProvider]: never;
+	[ErrorCode.InvalidProviderConfig]: InvalidProviderConfigErrorExtensions;
+	[ErrorCode.InvalidQuery]: InvalidQueryErrorExtensions;
+	[ErrorCode.InvalidToken]: never;
+	[ErrorCode.LimitExceeded]: never;
+	[ErrorCode.MethodNotAllowed]: MethodNotAllowedErrorExtensions;
+	[ErrorCode.NotNullViolation]: NotNullViolationErrorExtensions;
+	[ErrorCode.OutOfDate]: never;
+	[ErrorCode.RangeNotSatisfiable]: RangeNotSatisfiableErrorExtensions;
+	[ErrorCode.RecordNotUnique]: RecordNotUniqueErrorExtensions;
+	[ErrorCode.RequestsExceeded]: HitRateLimitErrorExtensions;
+	[ErrorCode.RouteNotFound]: RouteNotFoundErrorExtensions;
+	[ErrorCode.ServiceUnavailable]: ServiceUnavailableErrorExtensions;
+	[ErrorCode.TokenExpired]: never;
+	[ErrorCode.UnexpectedResponse]: never;
+	[ErrorCode.UnprocessableContent]: UnprocessableContentErrorExtensions;
+	[ErrorCode.UnsupportedMediaType]: UnsupportedMediaTypeErrorExtensions;
+	[ErrorCode.UserSuspended]: never;
+	[ErrorCode.ValueOutOfRange]: ValueOutOfRangeErrorExtensions;
+	[ErrorCode.ValueTooLong]: ValueTooLongErrorExtensions;
+};
+
+/** Verify all error codes are covered in the map. */
+type Extends<T, U extends T> = U;
+
+/** Map error codes to error extensions. */
+export type ExtensionsMap = Extends<Record<(typeof ErrorCode)[keyof typeof ErrorCode], any>, Map>;
```

Comment:
To make sure we always have all error codes covered by mapping them to `never` we can fallback any undefined errors like so

```suggestion
/** Make sure all error codes are covered in the map. */
export type ExtensionsMap = {
	[code in ErrorCode]: code extends keyof Map
		? Map[code]
		: never;
};
```

Reply:
The idea was to have a built-in constraint which ensures we have covered all errors, e.g. in case a new error should be introduced and it should be forgotten to update the map:
<img width="940" src="https://github.com/directus/directus/assets/5363448/c9d6041c-7ff6-4892-8570-9ed12def22ce">

But might be a bit overkill, so I'm fine with your suggestion, too. What do you prefer?

Reply:
After thinking about it again: Let's not over-complicate it, your suggestions feels more lightweight 👍

Reply:
> The idea was to have a built-in constraint which ensures we have covered all errors

I do agree that would be great and we should definitely aim to have full coverage there. However I do think we should then not allow `never` as escape hatch and actually define all those errors 😄